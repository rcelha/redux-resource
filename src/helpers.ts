import { get } from 'lodash';
import { StructuredResource } from './reducers/types';
import { INITIAL_RESOURCE } from './reducers/initial-resource';

/**
 * Get a resource from the store
 *
 * It works with single resource or lists, depending on
 * whether you sent and `id` or not
 */
export function getFrom(
  state: { resources: object },
  resource: string,
  id?: string | number
): StructuredResource {
  let uri;
  if (id) {
    id = id.toString();
    uri = `byId.${resource}.${id}`;
  } else {
    uri = `list.${resource}`;
  }

  return get(state.resources, uri, { ...INITIAL_RESOURCE });
}

/**
 * Get one or more queries
 *
 * When there is more than one query, it combines they in a singe StructuredResource,
 * appending the data, and merging the metadata
 */
export function getQueries(
  state: { resources: object },
  resource: string,
  queries: [object] = [{}]
): StructuredResource {
  // return get(state.resources, `query.${resource}`, { ...INITIAL_RESOURCE });
  const res = get(state.resources, `query.${resource}`, {});
  return queries.reduce(
    (acc: StructuredResource, query: object) => {
      const queryKey = JSON.stringify(query);
      const currentQuery = get(res, queryKey, {
        ...INITIAL_RESOURCE,
      }) as StructuredResource;
      const data = (acc.data as []) || [];
      const newData = (currentQuery.data as []) || [];
      acc = { ...acc, ...currentQuery };
      acc.data = [...data, ...newData];
      return acc;
    },
    { ...INITIAL_RESOURCE }
  );
}
