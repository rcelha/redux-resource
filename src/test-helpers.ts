import { applyMiddleware, createStore } from 'redux';
import createSagaMiddleware from 'redux-saga';
import { keyBy } from 'lodash';
import { saga } from './index';
import { reducer } from './reducers';

export const sleep = (milliseconds: number) => {
  return new Promise(resolve => setTimeout(resolve, milliseconds));
};

export const setupStore = () => {
  const sagaMiddleware = createSagaMiddleware();
  const store = createStore(reducer, applyMiddleware(sagaMiddleware));
  sagaMiddleware.run(saga);
  return store;
};

export const service = {
  fetchUser({ id }: { id: number }) {
    const users = service.fetchUsers();
    const data = keyBy(users.data, 'id')[id];
    return { data };
  },

  fetchUsers({ p = 0 } = {}) {
    const data = [
      [{ id: 1, name: 'Rodrigo' }, { id: 2, name: 'Fernanda' }],
      [{ id: 11, name: 'Cherry' }, { id: 12, name: 'Tyrion' }],
      [{ id: 21, name: 'Liz' }, { id: 22, name: 'Edo' }],
    ];

    return {
      data: data[p],
      meta: { total: 2 },
    };
  },

  postUser({ expand = false } = {}) {
    const data = expand
      ? { id: 100, name: 'New User From Service' }
      : { id: 999 };
    return { data };
  },
};
