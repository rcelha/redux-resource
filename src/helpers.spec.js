import { getFrom, getQueries } from './helpers';
import { setupStore } from './test-helpers';

describe('helpers', () => {
  describe('getFrom', () => {
    it('should return an non initialized resource', () => {
      const state = { resources: { byId: {}, list: {} } };
      const res = getFrom(state, 'users');
      expect(res).toMatchObject({
        loading: false,
        cached: false,
        initialized: false,
        data: null,
        meta: null,
        error: null,
      });
    });

    it('should return an non initialized single resource', () => {
      const state = { resources: { byId: {}, list: {} } };
      const res = getFrom(state, 'users', 1);
      expect(res).toMatchObject({
        loading: false,
        cached: false,
        initialized: false,
        data: null,
        meta: null,
        error: null,
      });
    });

    it('should return an initialized single resource', () => {
      const state = {
        resources: {
          byId: { users: { '1': { data: { id: 1, name: 'Test' } } } },
          list: {},
        },
      };
      const res = getFrom(state, 'users', 1);
      expect(res).toMatchObject({
        data: { id: 1, name: 'Test' },
      });
    });

    it('should return an initialized list of resources', () => {
      const state = {
        resources: {
          byId: {},
          list: { users: { data: [{ id: 1, name: 'Test' }] } },
        },
      };
      const res = getFrom(state, 'users');
      expect(res).toMatchObject({
        data: [{ id: 1, name: 'Test' }],
      });
    });
  });

  describe('getQueries', () => {
    const state = {
      resources: {
        byId: {},
        list: {},
        query: {
          users: {
            '{"p":1}': {
              cached: false,
              data: [
                {
                  id: 11,
                  name: 'Cherry',
                },
                {
                  id: 12,
                  name: 'Tyrion',
                },
              ],
              error: null,
              initialized: true,
              loading: false,
              meta: {},
            },
            '{"p":2}': {
              cached: false,
              data: [
                {
                  id: 21,
                  name: 'Liz',
                },
                {
                  id: 22,
                  name: 'Edo',
                },
              ],
              error: null,
              initialized: true,
              loading: false,
              meta: {},
            },
            '{}': {
              cached: false,
              data: [
                {
                  id: 1,
                  name: 'Rodrigo',
                },
                {
                  id: 2,
                  name: 'Fernanda',
                },
              ],
              error: null,
              initialized: true,
              loading: false,
              meta: {},
            },
          },
        },
      },
    };

    it('should extract a single query', () => {
      const queries = getQueries(state, 'users');
      expect(queries).toMatchSnapshot();
    });
    it('should combine multiple queries', () => {
      const queries = getQueries(state, 'users', [{}, { p: 2 }]);
      expect(queries).toMatchSnapshot();
    });
  });
});
