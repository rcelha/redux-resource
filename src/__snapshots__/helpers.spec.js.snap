// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`helpers getQueries should combine multiple queries 1`] = `
Object {
  "cached": false,
  "data": Array [
    Object {
      "id": 1,
      "name": "Rodrigo",
    },
    Object {
      "id": 2,
      "name": "Fernanda",
    },
    Object {
      "id": 21,
      "name": "Liz",
    },
    Object {
      "id": 22,
      "name": "Edo",
    },
  ],
  "error": null,
  "initialized": true,
  "loading": false,
  "meta": Object {},
}
`;

exports[`helpers getQueries should extract a single query 1`] = `
Object {
  "cached": false,
  "data": Array [
    Object {
      "id": 1,
      "name": "Rodrigo",
    },
    Object {
      "id": 2,
      "name": "Fernanda",
    },
  ],
  "error": null,
  "initialized": true,
  "loading": false,
  "meta": Object {},
}
`;
